sklearn/preprocessing/_csr_polynomial_expansion.pyx

```diff
@@ -0,0 +1,159 @@
+# cython: cdivision=True
+# cython: boundscheck=False
+# cython: wraparound=False
+
+# Author: Andrew nystrom <awnystrom@gmail.com>
+
+from scipy.sparse import csr_matrix
+from numpy cimport ndarray
+cimport numpy as np
+
+ctypedef np.int32_t INDEX_T
+
+ctypedef fused DATA_T:
+    np.float32_t
+    np.float64_t
+    np.int32_t
+    np.int64_t
+
+
+cdef inline INDEX_T _deg2_column(INDEX_T d, INDEX_T i, INDEX_T j,
+                                 INDEX_T interaction_only) nogil:
+    """Compute the index of the column for a degree 2 expansion
+
+    d is the dimensionality of the input data, i and j are the indices
+    for the columns involved in the expansion.
+    """
+    if interaction_only:
+        return d * i - (i**2 + 3 * i) / 2 - 1 + j
+    else:
+        return d * i - (i**2 + i) / 2 + j
+
+
+cdef inline INDEX_T _deg3_column(INDEX_T d, INDEX_T i, INDEX_T j, INDEX_T k,
+                                 INDEX_T interaction_only) nogil:
+    """Compute the index of the column for a degree 3 expansion
+
+    d is the dimensionality of the input data, i, j and k are the indices
+    for the columns involved in the expansion.
+    """
+    if interaction_only:
+        return ((3 * d**2 * i - 3 * d * i**2 + i**3
+                 + 11 * i - 3 * j**2 - 9 * j) / 6
+                + i**2 - 2 * d * i + d * j - d + k)
+    else:
+        return ((3 * d**2 * i - 3 * d * i**2 + i ** 3 - i
+                 - 3 * j**2 - 3 * j) / 6
+                + d * j + k)
+
+
+def _csr_polynomial_expansion(ndarray[DATA_T, ndim=1] data,
+                              ndarray[INDEX_T, ndim=1] indices,
+                              ndarray[INDEX_T, ndim=1] indptr,
+                              int has_sorted_indices, INDEX_T d,
```

Comment:
I just noticed that I left an unused parameter in my refactoring. I think this `has_sorted_indices` flag  is no longer required. The `csr_matrix` constructor we now use should automatically set this attribute I think.